lsp/lsp-harness/src/lib.rs

```diff
@@ -0,0 +1,137 @@
+mod jsonrpc;
+mod output;
+
+pub use jsonrpc::{ResponseHandle, Server};
+use log::error;
+use lsp_types::{
+    GotoDefinitionParams, Location, Position, Range, TextDocumentIdentifier,
+    TextDocumentPositionParams, Url,
+};
+pub use output::LspDebug;
+
+/// A text fixture consists of multiple files (each labelled with a filename)
+/// and multiple requests to run on those files.
+///
+/// The test is executed by loading the files into the language server and then
+/// executing the requests one-by-one. The responses are printed and compared
+/// against a reference response.
+pub struct TestFixture {
+    pub files: Vec<TestFile>,
+    pub reqs: Vec<Request>,
+}
+
+impl TestFixture {
+    pub fn parse(s: &str) -> Option<Self> {
+        <Self as Parse>::parse(s)
+    }
+}
+
+pub struct TestFile {
+    pub uri: Url,
+    pub contents: String,
+}
+
+/// A subset of LSP requests that our harness supports.
+pub enum Request {
+    GotoDefinition(GotoDefinitionParams),
+}
+
+/// A private FromStr, with simpler error handling and (because it's
+/// private) the ability to impl it for external types.
+trait Parse: Sized {
+    fn parse(s: &str) -> Option<Self>;
+}
```

Comment:
I started out using `FromStr` and the newtype wrappers, but the custom trait ended up being slightly less boilerplate because converting from `Result<T, ()>` to `Option<T>` is a little nicer than going from `Option<T>` (or sometimes `Result<T, IntParseError>` to `Result<T, ()>`). With the `FromStr`, I had a lot of `map_err(|_| ())?`...

I could go either way, though; it wasn't a huge difference.

Reply:
I was mostly curious, but if you tried both, you're probably in the best position to know which one is better, so do as you think is best